crates/libs/core/src/types/client/health.rs

```diff
@@ -0,0 +1,138 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+use mssf_com::FabricTypes::{
+    FABRIC_HEALTH_REPORT, FABRIC_HEALTH_REPORT_KIND_APPLICATION, FABRIC_HEALTH_REPORT_KIND_CLUSTER,
+    FABRIC_HEALTH_REPORT_KIND_DEPLOYED_APPLICATION,
+    FABRIC_HEALTH_REPORT_KIND_DEPLOYED_SERVICE_PACKAGE, FABRIC_HEALTH_REPORT_KIND_INVALID,
+    FABRIC_HEALTH_REPORT_KIND_NODE, FABRIC_HEALTH_REPORT_KIND_PARTITION,
+    FABRIC_HEALTH_REPORT_KIND_SERVICE, FABRIC_HEALTH_REPORT_KIND_STATEFUL_SERVICE_REPLICA,
+    FABRIC_HEALTH_REPORT_KIND_STATELESS_SERVICE_INSTANCE,
+};
+
+/// Wrapper of FABRIC_HEALTH_REPORT
+pub enum HealthReport {
+    Invalid,
+    StatefulServiceReplica(StatefulServiceReplicaHealthReport),
+    StatelessServiceInstance(StatelessServiceInstanceHealthReport),
+    Partition(PartitionHealthReport),
+    Node(NodeHealthReport),
+    Service(ServiceHealthReport),
+    Application(ApplicationHealthReport),
+    DeployedApplication(DeployedApplicationHealthReport),
+    DeployedServicePackage(DeployedServicePackageHealthReport),
+    Cluster(ClusterHealthReport),
+}
+
+// TODO: Implement the conversion functions for the HealthReport enum properly.
+impl From<&FABRIC_HEALTH_REPORT> for HealthReport {
+    fn from(value: &FABRIC_HEALTH_REPORT) -> Self {
+        match value.Kind {
+            FABRIC_HEALTH_REPORT_KIND_STATEFUL_SERVICE_REPLICA => {
+                HealthReport::StatefulServiceReplica(StatefulServiceReplicaHealthReport)
+            }
+            FABRIC_HEALTH_REPORT_KIND_STATELESS_SERVICE_INSTANCE => {
+                HealthReport::StatelessServiceInstance(StatelessServiceInstanceHealthReport)
+            }
+            FABRIC_HEALTH_REPORT_KIND_PARTITION => HealthReport::Partition(PartitionHealthReport),
+            FABRIC_HEALTH_REPORT_KIND_NODE => HealthReport::Node(NodeHealthReport),
+            FABRIC_HEALTH_REPORT_KIND_SERVICE => HealthReport::Service(ServiceHealthReport),
+            FABRIC_HEALTH_REPORT_KIND_APPLICATION => {
+                HealthReport::Application(ApplicationHealthReport)
+            }
+            FABRIC_HEALTH_REPORT_KIND_DEPLOYED_APPLICATION => {
+                HealthReport::DeployedApplication(DeployedApplicationHealthReport)
+            }
+            FABRIC_HEALTH_REPORT_KIND_DEPLOYED_SERVICE_PACKAGE => {
+                HealthReport::DeployedServicePackage(DeployedServicePackageHealthReport)
+            }
+            FABRIC_HEALTH_REPORT_KIND_CLUSTER => HealthReport::Cluster(ClusterHealthReport),
+            _ => HealthReport::Invalid,
+        }
+    }
+}
+
+impl From<&HealthReport> for FABRIC_HEALTH_REPORT {
+    fn from(value: &HealthReport) -> Self {
+        match value {
```

Comment:
does it need to be *mut std::ffi_cvoid over *const? bad smell

Reply:
For core-com conversions, this is unavoidable, I think.

Reply:
It seems like the this does not work - segfault from local testing. Need to look into on how do properly do core-com-core conversions

Reply:
Yes, you are right. The following type conversions are incorrect. I need to implement From trait for each individual type here. Actually, GitHub copilot generated these for me :-)